doc/developer/design/20240117_decoupled_storage_controller.md

```diff
@@ -0,0 +1,92 @@
+# Platform v2: Decouple Personalities of the Storage Controller
+
+> [!NOTE]
+> This is more a proto-design doc right now, laying out the context and basics
+> of what I want.
+
+## Context
+
+As part of the platform v2 work (specifically use-case isolation) we want to
+develop a scalable and isolated serving layer that is made up of multiple
+processes that interact with distributed primitives at the moments where
+coordination is required.
+
+The way the compute controller and storage controller work currently will not
+work well for that, because they're coupled together inside one process.
+
+## Goals
+
+- Make "storage controller" shareable, give each compute controller their own
+  instance. This actually means we want to give each compute controller a
+  CollectionController, but you have to read below first.
+- Tweeze appart responsibilities of the current storage controller.
+- Use persist directly (helped by persist pubsub) to learn about uppers of
+  collections and drive their since forward (based on the since policy).
+
+## Non-Goals
+
+- Actually separate things out into separate processes.
+
+## Overview
+
+- We want to to have Coordinator-like things running in different processes. We
+  want each cluster's (compute) controller to be separate from other
+  controllers, running in separate processes.
+- Currently, the compute controller is given a "mutable reference" to the
+  storage controller, when it needs to do any work.
+- The storage controller has come to be responsible for 3 things that (IMO) can
+  and should be separated out:
+  - CollectionController: holds on to since handles and tells a compute
+    controller at what time collections are readable, when asked. Allows
+    installing read holds.
+  - TablesController: facilitates writes to tables. Could also be called
+    TxnController.
+  - StorageController: manages storage-flavored computation running on a
+    cluster. Think sources and sinks.
+
+Sketch of what I think we need:
+
+- The compute controller really only needs a CollectionController. It would
+  have a much reduced interface, compared to current StorageController. Greatly
+  reducing surface area/coupling between the two.
+- We can revive an old idea where we use persist directly to learn about the
+  upper of collections: the CollectionController would not rely on feedback
+  from the cluster to learn about uppers. It would use persist, and drive it's
```

Comment:
There is more information that flows from active ingestion dataflows into the controller. It sounds like the compute controller will be close to the cluster that it's managing which makes sense and I think the same should be true for the storage controller for the ingestions that *it* manages.

I could see shareable lightweight storage controller-like things becoming a thing that is given to other processes that want **read** access to collections but anything that writes to a collection I think should continue to operate under the current model. Is something incompatible with this mode?

Reply:
That is what I had in mind, yes! What threw me off, and still seems confusing then, is why does the `CollectionController` need to learn about uppers of collections. It sounds like at most it would need to attempt to grab a read handle at some requested frontier

Reply:
When processing queries, the adapter/coordinator wants to know the upper of collections. And, IMO, the CollectionController would hold since handles for collections, to keep them readable, and would drive that forward based on the upper. Same as what the storage controller does now, but in a more distributed way because you can't have a full storage controller in all the processes that want to respond to queries.

Reply:
> There is more information that flows from active ingestion dataflows into the controller.

I think this is some missing context that I don't yet have about the storage controller. Do either of you have some examples of these?

> but anything that writes to a collection I think should continue to operate under the current model

Does "the current model" mean a singleton-per-env storage controller or (the rest of this thread read to me) more like singleton-per-cluster? If the latter, where do tables which in a Use Case Isolation world might be written to from any cluster, fit into it? Ditto the various storage introspection sources.

Reply:
> I think this is some missing context that I don't yet have about the storage controller. Do either of you have some examples of these?

I'd say it's pretty much `StorageResponse` ([code](https://github.com/MaterializeInc/materialize/blob/8b720344dc20645b94c9f09ddd25e7f3c8b59ab6/src/storage-client/src/client.rs#L402)): so also statistics updates (think ingested messages and the like), status updates, and `DroppedIds` (which are an implementation detail of the cluster-local `StorageController`).

> Does "the current model" mean a singleton-per-env storage controller or (the rest of this thread read to me) more like singleton-per-cluster? If the latter, where do tables which in a Use Case Isolation world might be written to from any cluster, fit into it?

There has to be a (new-style, slimmed down, as per this doc) `StorageController` per cluster, similar to how the `ComputeController` becomes a per-cluster thing. And `TxnController`/`TableController` (which, yes, I only briefly mention in the doc) would be similar to `CollectionController`. Anyone who needs to do table writes can hold one of these and they're "shareable"/do coordination via the distributed things they hold: since handles for the `CollectionController` (which defers distributed coordination to persist), and the persist-txn system for `TxnController`, which defers coordination to the txn system.

> Ditto the various storage introspection sources.

`StorageController` (the new-style, per-cluster instances) have to collaborate in writing these down.